tests/queries/tests.py

```diff
@@ -2572,7 +2573,7 @@ def test_evaluated_queryset_as_argument(self):
         list(n_list)
         # Make one of cached results unpickable.
         n_list._result_cache[0].lock = Lock()
-        with self.assertRaises(TypeError):
+        with self.assertRaises(pickle.PicklingError if PYPY else TypeError):
             pickle.dumps(n_list)
```

Comment:
Here's a simple reproducer in Python 3.10:

```pycon
>>> import pickle, threading
>>> lock = threading.Lock()
>>> pickle.dumps(lock)  # Using C
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: cannot pickle '_thread.lock' object
>>> pickle._dumps(lock)  # Using Python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.10/pickle.py", line 1757, in _dumps
    buffer_callback=buffer_callback).dump(obj)
  File "/usr/lib/python3.10/pickle.py", line 487, in dump
    self.save(obj)
  File "/usr/lib/python3.10/pickle.py", line 578, in save
    rv = reduce(self.proto)
TypeError: cannot pickle '_thread.lock' object
```

And here's the same with PyPy:

```pycon
>>>> import pickle, threading
>>>> lock = threading.Lock()
>>>> pickle.dumps(lock)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/pypy3/lib/pypy3.10/pickle.py", line 1769, in _dumps
    _Pickler(f, protocol, fix_imports=fix_imports,
  File "/opt/pypy3/lib/pypy3.10/pickle.py", line 500, in dump
    self.save(obj)
  File "/opt/pypy3/lib/pypy3.10/pickle.py", line 616, in save
    self.save_reduce(obj=obj, *rv)
  File "/opt/pypy3/lib/pypy3.10/pickle.py", line 700, in save_reduce
    save(cls)
  File "/opt/pypy3/lib/pypy3.10/pickle.py", line 573, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/opt/pypy3/lib/pypy3.10/pickle.py", line 1142, in save_type
    return self.save_global(obj)
  File "/opt/pypy3/lib/pypy3.10/pickle.py", line 1084, in save_global
    raise PicklingError(
pickle.PicklingError: Can't pickle <class '_thread.lock'>: it's not found as _thread.lock
```

Note the difference - it would seem that we're falling back to one of the non-pickleable cases raised out of [`Objects/typeobject.c`](https://github.com/python/cpython/blob/1b8aa7aafd2ac07836777707fd9ba5ffb353eb0c/Objects/typeobject.c) for CPython, but we don't seem to hit some equivalent in PyPy, but something else. This isn't therefore an issue with the `pickle` module, but some deeper down implementation difference.

Interestingly, I think the underlying issue here is that the locks show themselves as being `_thread.lock` which just isn't a thing!

CPython:

```pycon
>>> from _thread import lock
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'lock' from '_thread' (unknown location)
```

PyPy:

```pycon
>>>> from _thread import lock
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'lock' from '_thread' (unknown location)
```

Reply:
Can any of you open an issue in Python and share your findings?

Reply:
I think this is something to be fixed in PyPy, or  maybe we need to change the test. Ultimately we just want something unpicklable here.

Yes, `_thread.lock` isn't a thing, but that's to be expected - it's been this way forever and this is a very low level implementation thing that isn't meant to be used directly.

The discrepancy seems to be either due to the way that PyPy handles or implements the low level thread stuff which is not part of the standard library, or the way that it implemented the equivalent of stuff from `typeobject.c` which is where the `TypeError` is coming from on CPython.

Reply:
For further consideration related to this, I get the following if I attempt to pickle `dict_items`...

CPython:

```pycon
>>> import pickle
>>> pickle.dumps({}.items())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: cannot pickle 'dict_items' object
```

PyPy:

```pycon
>>>> import pickle
>>>> pickle.dumps({}.items())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/pypy3/lib/pypy3.10/pickle.py", line 1769, in _dumps
    _Pickler(f, protocol, fix_imports=fix_imports,
  File "/opt/pypy3/lib/pypy3.10/pickle.py", line 500, in dump
    self.save(obj)
  File "/opt/pypy3/lib/pypy3.10/pickle.py", line 616, in save
    self.save_reduce(obj=obj, *rv)
  File "/opt/pypy3/lib/pypy3.10/pickle.py", line 700, in save_reduce
    save(cls)
  File "/opt/pypy3/lib/pypy3.10/pickle.py", line 573, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/opt/pypy3/lib/pypy3.10/pickle.py", line 1142, in save_type
    return self.save_global(obj)
  File "/opt/pypy3/lib/pypy3.10/pickle.py", line 1084, in save_global
    raise PicklingError(
pickle.PicklingError: Can't pickle <class 'dict_items'>: it's not found as builtins.dict_items
```

In CPython, when we look at `static PyObject *reduce_newobj(PyObject *obj)` in `Objects/typeobject.c` we can see:

```c
    if (Py_TYPE(obj)->tp_new == NULL) {
        PyErr_Format(PyExc_TypeError,
                     "cannot pickle '%.200s' object",
                     Py_TYPE(obj)->tp_name);
        return NULL;
    }
```

And looking at the implementation for `dict_items` in `Objects/dictobject.c` we can see that `PyTypeObject PyDictItems_Type` does not implement `tp_new`. In CPython we also have the following in `static PyObject *
type_call(PyTypeObject *type, PyObject *args, PyObject *kwds)`:

```c
    if (type->tp_new == NULL) {
        _PyErr_Format(tstate, PyExc_TypeError,
                      "cannot create '%s' instances", type->tp_name);
        return NULL;
    }
```

And so, in CPython:

```pycon
>>> from threading import Lock
>>> type(Lock())()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: cannot create '_thread.lock' instances
>>> type({}.items())({"a": 1})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: cannot create 'dict_items' instances
```

Compared to PyPy:

```pycon
>>>> from threading import Lock
>>>> type(Lock())()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: object.__new__(lock) is not safe, use lock.__new__()
>>>> type({}.items())({"a": 1})
dict_items([('a', 1)])
```

@cfbolz @mattip Any ideas on how to proceed here? I don't think that CPython is wrong - these are just types that cannot be created this way, yet PyPy is able to do so in some cases. I had a look in `pypy/objspace/std/objectobject.py`, but don't really know where to begin.

@felixxm Maybe, as mentioned above, we should just use a custom thing, e.g.

CPython:

```pycon
>>> class Unpickleable:
...     def __reduce__(self):
...         raise Exception("Cannot pickle.")
... 
>>> import pickle
>>> pickle.dumps(Unpickleable())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in __reduce__
Exception: Cannot pickle.
```

PyPy:

```pycon
>>>> class Unpickleable:
....     def __reduce__(self):
....         raise Exception("Cannot pickle.")
....         
>>>> import pickle
>>>> pickle.dumps(Unpickleable())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/pypy3/lib/pypy3.10/pickle.py", line 1769, in _dumps
    _Pickler(f, protocol, fix_imports=fix_imports,
  File "/opt/pypy3/lib/pypy3.10/pickle.py", line 500, in dump
    self.save(obj)
  File "/opt/pypy3/lib/pypy3.10/pickle.py", line 591, in save
    rv = reduce(self.proto)
  File "<stdin>", line 3, in __reduce__
Exception: Cannot pickle.
```

Looks like it does the job!